crates/aptos-jwk-consensus/src/jwk_manager/mod.rs

```diff
@@ -0,0 +1,408 @@
+
+    /// The JWK consensus states of all the issuers.
+    states_by_issuer: HashMap<Issuer, PerProviderState>,
+
+    /// Whether a CLOSE command has been received.
+    stopped: bool,
+
+    qc_update_tx: Option<aptos_channel::Sender<(), QuorumCertifiedUpdate>>,
+    jwk_observers: Vec<JWKObserver>,
+}
+
+impl JWKManager {
+    pub fn new(
+        consensus_key: Arc<PrivateKey>,
+        my_addr: AccountAddress,
+        epoch_state: Arc<EpochState>,
+        certified_update_producer: Arc<dyn CertifiedUpdateProducer>,
+        vtxn_pool: VTxnPoolState,
+    ) -> Self {
+        Self {
+            consensus_key,
+            my_addr,
+            epoch_state,
+            certified_update_producer,
+            vtxn_pool,
+            states_by_issuer: HashMap::default(),
+            stopped: false,
+            qc_update_tx: None,
+            jwk_observers: vec![],
+        }
+    }
+
+    pub async fn run(
+        mut self,
+        oidc_providers: Option<SupportedOIDCProviders>,
+        observed_jwks: Option<ObservedJWKs>,
+        mut jwk_updated_rx: aptos_channel::Receiver<(), ObservedJWKsUpdated>,
+        mut rpc_req_rx: aptos_channel::Receiver<(), (AccountAddress, IncomingRpcRequest)>,
+        close_rx: oneshot::Receiver<oneshot::Sender<()>>,
+    ) {
+        self.reset_with_on_chain_state(observed_jwks.unwrap_or_default().into_providers_jwks())
+            .unwrap();
+        let (qc_update_tx, mut qc_update_rx) = aptos_channel::new(QueueStyle::FIFO, 100, None);
+        self.qc_update_tx = Some(qc_update_tx);
+
+        let (local_observation_tx, mut local_observation_rx) =
+            aptos_channel::new(QueueStyle::KLAST, 100, None);
+
+        self.jwk_observers = oidc_providers
+            .unwrap_or_default()
+            .into_provider_vec()
+            .into_iter()
+            .map(|provider| {
+                JWKObserver::spawn(
+                    self.my_addr,
+                    provider.name.clone(),
+                    provider.config_url.clone(),
+                    Duration::from_secs(10),
+                    local_observation_tx.clone(),
+                )
+            })
+            .collect();
+
+        let mut close_rx = close_rx.into_stream();
+
+        while !self.stopped {
+            let handle_result = tokio::select! {
+                jwk_updated = jwk_updated_rx.select_next_some() => {
+                    let ObservedJWKsUpdated { jwks, .. } = jwk_updated;
+                    self.reset_with_on_chain_state(jwks)
+                },
+                (_sender, msg) = rpc_req_rx.select_next_some() => {
+                    self.process_peer_request(msg)
+                },
+                qc_update = qc_update_rx.select_next_some() => {
+                    self.process_quorum_certified_update(qc_update)
+                },
+                (issuer, jwks) = local_observation_rx.select_next_some() => {
+                    let jwks = jwks.into_iter().map(JWKMoveStruct::from).collect();
```

Comment:
This is deterministic right? every validator should have same `jwks`?

Reply:
yes, sorted in #11860